Style heading pattern's typography

davidcmoulton opened the pull request:
No web fonts yet.

davidcmoulton committed:
Add sass capability

Using Dart sass to get the latest features

davidcmoulton committed:
Migrate settings as sass modules

davidcmoulton committed:
Not needed (yet)

davidcmoulton committed:
Add sass-rem dependency

davidcmoulton committed:
Add typography styling for heading pattern

Ports over enough of the Pattern Lab sass to enable the desired styling.

- takes advantage of Sass modules
- no web fonts yet

davidcmoulton committed:
Use sass preset for Storybook

davidcmoulton committed:
lint

davidcmoulton committed:
Validate heading level

davidcmoulton commented on src/core-styling/functions/_validation.scss:
```diff
@@ -0,0 +1,9 @@
+$_is-test: false !default;
+
+@function error($message, $capture: $_is-test) {
```
Previously was `_error`, but that enforces a privacy when `@use`d that we don't want in this case.

davidcmoulton commented on src/core-styling/functions/_validation.scss:
```diff
@@ -0,0 +1,9 @@
+$_is-test: false !default;
```
Should not be private.

davidcmoulton commented on src/core-styling/mixins/_typography.scss:
```diff
@@ -0,0 +1,24 @@
+}
+
+@mixin covert-link($hover-color: color.$primary-dark) {
+  @include utilities.inherit-all($ensure: color text-decoration) {
```
Still not sure whether `inherit-all` is too blunt an instrument. Will leave it for now, to reconsider later.

thewilkybarkid replied:
I think it proved to be, and should probably not carry it over.

davidcmoulton commented on src/patterns/molecules/heading/heading.scss:
```diff
@@ -0,0 +1,81 @@
+  $scale: _get-scale-point($heading-level);
+  $line-height: _get-line-height($heading-level);
+  @include _base-typography();
+  @include typography.set-font-size-and-line-height(scale($scale), $line-height);
```
Missing `scale`.

davidcmoulton committed:
Not private

davidcmoulton committed:
Add modular scale

davidcmoulton commented on src/core-styling/mixins/_scale.scss:
```diff
@@ -0,0 +1,6 @@
+@use "../settings/font";
+@use "../../../node_modules/modularscale-sass/stylesheets/modularscale";
```
Not that happy about reaching down directly into `node_modules`, but equally not happy about tying the sass implementation to Webpack by [using the `~` prefix](https://webpack.js.org/loaders/sass-loader/#resolving-import-at-rules) either 🤔.

thewilkybarkid replied:
Since the path won't work if this were turned into an NPM module, probably better to use the Webpack syntax for now...

davidcmoulton replied:
True enough 👍

davidcmoulton commented on src/patterns/molecules/heading/heading.scss:
```diff
@@ -0,0 +1,82 @@
+@use "../../../core-styling/mixins/scale";
+@use "../../../core-styling/mixins/validation";
+
+$_h1-line-height: baselinegrid.$space-medium;
```
Alternative would be to make these configurable with an implementation like e.g.: 
```
$h1-line-height: baselinegrid.$space-medium !default;
```
However this is outside the scope of configurability we're allowing, so unless there're good reasons not to, will keep them as private, and soley defined by the baseline grid.

davidcmoulton commented on src/core-styling/mixins/_types.scss:
```diff
@@ -0,0 +1,44 @@
+@use "../../../node_modules/sass-rem/rem";
+
+$rem-baseline: font.$size;
+$rem-fallback: true;
```
Can safely be `false` now.

thewilkybarkid replied:
Would simplify things, and means we don't need the override either.

thewilkybarkid commented on src/patterns/molecules/heading/heading.scss:
```diff
@@ -0,0 +1,82 @@
+@mixin _typography($heading-level) {
+  $permitted: (level1, level2, level3, level4, level5, level6);
+  @if not list.index($permitted, $heading-level) {
+    @include validation.error("Invalid heading level supplied: #{$heading-level}");
```
Since this is private, this isn't needed (could never test for it).

davidcmoulton replied:
Yeah. I was getting carried away with _validateAllTheInputz!_

thewilkybarkid commented on src/core-styling/settings/_font.scss:
```diff
@@ -0,0 +1,5 @@
+$letterspacing-label: 0.5px !default;
```
Not needed

thewilkybarkid commented on src/core-styling/settings/_color.scss:
```diff
@@ -0,0 +1,21 @@
+$success: rgb(98, 159, 67) !default;
+$success_dark: rgb(86, 144, 55) !default;
+$attention: rgb(207, 12, 78) !default;
+$warning: rgb(230, 81, 0) !default;
```
I recommend removing all the colours not needed in favour of some rethinking (I wouldn't be able to tell you what `text-ui_background_hue__reverse` means).

davidcmoulton replied:
Agree. The configuration requirements, combined with the new flexibility provided by Sass modules, provide a good opportunity.

thewilkybarkid commented on src/core-styling/mixins/_scale.scss:
```diff
@@ -0,0 +1,6 @@
+@use "../../../node_modules/modularscale-sass/stylesheets/modularscale";
+
+@function scale($value: 0) {
+  @return round(modularscale.ms-function($value, font.$size, modularscale.$major-second));
```
I would recommend defining the base size and the scale together (one can be changed, the other can't).

davidcmoulton replied:
Yup. I'm working on an improvement around baseline and scale, based on the concepts we discussed yesterday.

thewilkybarkid commented on src/patterns/molecules/heading/heading.scss:
```diff
@@ -0,0 +1,82 @@
+  $line-height: _get-line-height($heading-level);
+  @include _base-typography();
+  @include typography.set-font-size-and-line-height(scale.scale($scale), $line-height);
+}
```
There's quite a lot of indirection here, would be easier to understand with a simpler implementation.

davidcmoulton replied:
Hopefully https://github.com/libero/storybook/pull/21/commits/6bf05199374456029a484486c0063036d13a8e11 makes it clearer 😅

davidcmoulton committed:
Configure Sass sourcemaps

davidcmoulton committed:
Improve baseline grid settings and usage

davidcmoulton committed:
Webpack path syntax for external sass dependencies

davidcmoulton committed:
Remove px fallback for rem

Browser support for rem is good enough not to need this anymore

thewilkybarkid commented on src/core-styling/settings/_baseline.scss:
```diff
@@ -0,0 +1,43 @@
+$unit-large: $_baseline-unit * $_factor-large;
+$unit-extra-large: $_baseline-unit * $_factor-extra-large;
+
+@function scale($value: 0) {
```
Thinking about the new modules, probably doesn't make sense to split functions, mixins, settings and utilities anymore. (So we'd have cases like this, with settings, functions etc.)

davidcmoulton replied:
I think so. Will need to be mindful of not blocking configuration where necessary, but now we're not looking to share cross-product sass variables in the way we previously were, it gives more room to manoeuvre here.

davidcmoulton committed:
Remove unneeded input validation

thewilkybarkid commented on src/core-styling/settings/_baseline.scss:
```diff
@@ -0,0 +1,43 @@
+//----------------------------------------------------------------------------------------------------------------------
+// Spacing values are determined based on the global baseline grid.
+//
+// It is not recommended that you modify these following variables (it can break your vertical rhythm).
```
If there's no `!default`, they can't be modified can they?

thewilkybarkid commented on src/core-styling/settings/_baseline.scss:
```diff
@@ -0,0 +1,43 @@
+
+$unit-extra-small: $_baseline-unit * $_factor-extra-small;
+$unit-small: $_baseline-unit * $_factor-small;
+$unit-smallish: $_baseline-unit * $_factor-smallish;
```
Time to revisit the names? 😅

davidcmoulton replied:
😅

thewilkybarkid commented on src/patterns/molecules/heading/heading.scss:
```diff
@@ -0,0 +1,76 @@
+  @include typography.set-font-size-and-line-height(baseline.scale($scale), $line-height);
+}
+
+h1 {
```
Bit of a different implementation, where the basic styling was done in a global normalization (so you didn't _have_ to use the heading pattern to have a `<h1>` etc).

Think it does make sense to try and do this, but makes we wonder about normalization. (Would we do the same with lists, for example?)

davidcmoulton replied:
I'm not sure yet that this is the final implementation. Going for encapsulation first and then to break things out where it makes sense to, hopeful to be able to discover a happy medium.

davidcmoulton replied:
I'm going to try breaking the heading typography into its own module. It can then be `@use`d in both the heading and normalisation stylesheets. A couple of other things I want to try too...

davidcmoulton committed:
Reduce indirection

davidcmoulton committed:
Fix H2 line height

davidcmoulton committed:
Not needed (for now)

davidcmoulton committed:
Remove obsolete comment

davidcmoulton committed:
Move file for clarity

thewilkybarkid committed:
Merge branch 'master' into configure-syling

davidcmoulton committed:
Not needed (yet)

davidcmoulton committed:
Remove as sledgehammer

davidcmoulton committed:
Development of sass modules idea

davidcmoulton committed:
Extract covert-link mixin

davidcmoulton committed:
Simplify heading typography calculation

davidcmoulton committed:
Only use the colours required for the current code

davidcmoulton committed:
Refactor (rename for clarity)

davidcmoulton committed:
Remove custom namespace

thewilkybarkid commented on .storybook/webpack.config.js:
```diff
@@ -1,17 +1,19 @@
-  });
+      ],
+    },
+  );
```
Looks like just a CS change?

thewilkybarkid commented on src/core-styling/_baseline.scss:
```diff
@@ -0,0 +1,43 @@
+
+//----------------------------------------------------------------------------------------------------------------------
+// Spacing values are determined based on the global baseline grid.
+//
```
Might want to be consistent on how to comment. 😅

thewilkybarkid commented on src/core-styling/_types.scss:
```diff
@@ -0,0 +1,7 @@
+  $rem-fallback: false
+);
+
+@forward "~sass-rem/rem";
```
This module should be called `rem`?

thewilkybarkid commented on src/core-styling/_baseline.scss:
```diff
@@ -0,0 +1,43 @@
+$unit-large: $_baseline-unit * $_factor-large;
+$unit-extra-large: $_baseline-unit * $_factor-extra-large;
+
+@function scale($value: 0) {
```
Not sure this should have a default?

thewilkybarkid commented on src/core-styling/_link.scss:
```diff
@@ -0,0 +1,12 @@
+  text-decoration: none;
+
+  &:hover {
+    color: color.$primary-dark;
```
```suggestion
    color: $hover-color;
```

thewilkybarkid commented on src/core-styling/_normalize.scss:
```diff
@@ -0,0 +1,6 @@
+*:after {
+  box-sizing: border-box;
+}
+
```
```suggestion
```

thewilkybarkid commented on src/core-styling/_typography-heading.scss:
```diff
@@ -0,0 +1,67 @@
+@forward "typography-global";
```
Don't think it should?

davidcmoulton replied:
👍 Old code. Removing.

thewilkybarkid commented on src/core-styling/_typography-heading.scss:
```diff
@@ -0,0 +1,67 @@
+
+@mixin _typography($scale-point, $line-height) {
+  font-family: $_font-family;
+  font-weight: 600;
```
Might as well make that a variable too.

thewilkybarkid commented on src/core-styling/_typography-heading.scss:
```diff
@@ -0,0 +1,67 @@
+  @include _typography(0, $_default_line-height);
+}
+
+@mixin _typography($scale-point, $line-height) {
```
Wonder about the order of the file (public vs private API, and selectors).

davidcmoulton replied:
Order matters for compilation: the mixins need to be defined before they're used in the selectors, and the local variables need to be defined before the mixin they're called from is executed.

Update: have changed the order as much as practicable within constraints.

thewilkybarkid commented on src/core-styling/_types.scss:
```diff
@@ -0,0 +1,7 @@
+  $rem-fallback: false
+);
+
+@forward "~sass-rem/rem";
```
```suggestion
@forward "~sass-rem/rem" show rem;
```

(Not tested; does this forward both the mixin and the function?)

davidcmoulton replied:
Don't like the ambiguity there. `@forward` removed in the refactor, see https://github.com/libero/storybook/pull/21/files#diff-cfeef31a60b46e4584db50a2f02af416

thewilkybarkid commented on src/core-styling/_typography-heading.scss:
```diff
@@ -0,0 +1,67 @@
+$_default_line-height: baseline.$unit-small;
+$_font-family: font.$secondary;
+
+@mixin h1-typography() {
```
Wondering whether this should just be `h1`? The module is specifically about typography, so `heading-typography.h1-typography` seems a bit redundant.

davidcmoulton committed:
Rename sass module for clarity

davidcmoulton committed:
Try to improve naming a bit

davidcmoulton committed:
Remove inappropriate default

davidcmoulton committed:
Improve how hover color is specified

davidcmoulton committed:
Whitespace

davidcmoulton committed:
Make heading font weight a variable

davidcmoulton committed:
Rename mixins to reduce repitition

davidcmoulton committed:
Order

davidcmoulton committed:
Revert unrelated code style change

davidcmoulton committed:
Consistent comment format within a file

davidcmoulton commented on src/core-styling/_baseline.scss:
```diff
@@ -0,0 +1,42 @@
+//----------------------------------------------------------------------------------------------------------------------
+// Spacing values are determined based on the global baseline grid.
+//
+$unit-extra-small: $_baseline-unit * $_factor-extra-small;
```
(Will supply a followup PR is to move to a 12px baseline unit, and replace the `$unit-[x]` variables with a function.)

thewilkybarkid committed:
Simplify import

thewilkybarkid committed:
CS fixes

thewilkybarkid committed:
Remove redundant variable

thewilkybarkid committed:
Remove single use variables

thewilkybarkid committed:
Fix calculation

thewilkybarkid committed:
Terminology

thewilkybarkid committed:
Avoid double namespace

thewilkybarkid committed:
Simplifications

thewilkybarkid committed:
Not needed

thewilkybarkid committed:
Make everything a module

thewilkybarkid committed:
Normalise basic typography

thewilkybarkid committed:
Rename sizes

thewilkybarkid commented on src/core-styling/_scale.scss:
```diff
@@ -0,0 +1,10 @@
+$base: font.$size;
+$ratio: modularscale.$major-second !default;
+
+@function at($point) {
```
Might need a better name.

thewilkybarkid committed:
Simpler name

thewilkybarkid reviewed (approved):
I'm made quite a few changes re the modularisation. Hopefully for the better, but please took take a look and revert if they don't make sense.

thewilkybarkid committed:
Simpler name

thewilkybarkid committed:
Plural

thewilkybarkid committed:
Meant to include

thewilkybarkid committed:
Use function

davidcmoulton committed:
Singular

davidcmoulton committed:
Refactor baseline variables

Includes change of naming strategy

davidcmoulton committed:
Consistent parameter names

davidcmoulton committed:
Not needed

davidcmoulton committed:
Define before use so it actually works

davidcmoulton committed:
Not typography

davidcmoulton committed:
Always use normalize first

davidcmoulton committed:
More precise name

davidcmoulton committed:
Use alphabetically (except normalize, which always comes first)

davidcmoulton committed:
Refactor to reality and halve the baseline unit

davidcmoulton committed:
Baseline unit of 6px addresses h2 line height gymnastics

davidcmoulton commented on src/shared-styles/_baseline.scss:
```diff
@@ -0,0 +1,38 @@
+  @return $_baseline-unit * $factor;
+}
+
+$size-1: _baseline-unit(1); // 6px
```
Could make these private and expose those currently required via names of convenience. It raises the question of what those names should be. Some options are:
1. keep `size` variables public and just use these. Not a bad option, and it means we don't have to keep coming up with sensible convenience names whenever a new `size` value needs making public
1. make `size` variables private and use the original names for convenience public variables or
1.  make `size` variables private and use newly suggested names for convenience public variables or
1.  make `size` variables private and use Reviewer's names for convenience public variables if they would work sensibly here too.

Right now options 1 and 4 look the best, with a slight preference for 1.

davidcmoulton replied:
**Update**: Decided to use a [function](https://github.com/libero/storybook/pull/21/files#diff-cec04b6d5a3012af00cfbdfa5a559686R15) to get the appropriate from the number of baseline units, comparable to the [`scale.at()` function](https://github.com/libero/storybook/pull/21/files#diff-e57318c7e4db838edcbdc32fb2979662R8)

davidcmoulton committed:
Access baseline size via function

davidcmoulton committed:
Order use statements

Default order is:
- normalize first if in a non-partial file
- sass modules, alphabetically
- external dependency modules, alphabetically
- local modules, alphabetically

N.B. This may occasionally change: when module A is used `with` a value  from module B, module B must be `use`d prior to module A, even though this breaks rules for alphabetical order.

davidcmoulton committed:
Further use of named parameters

davidcmoulton committed:
Case

thewilkybarkid commented on src/shared-styles/_baseline.scss:
```diff
@@ -12,27 +12,27 @@ $_baseline-unit: scale.$base * 0.375; // 6px
+$_size-17: calculateSizeFromUnits(17); // 102px
+$_size-18: calculateSizeFromUnits(18); // 108px
+$_size-19: calculateSizeFromUnits(19); // 114px
+$_size-20: calculateSizeFromUnits(20); // 120px
```
Aren't needed anymore?

thewilkybarkid commented on src/shared-styles/_baseline.scss:
```diff
@@ -0,0 +1,38 @@
+//----------------------------------------------------------------------------------------------------------------------
+// Spacing values determined by the global baseline grid.
+//
+@function calculate-size-from-units($factor) {
```
Hmm, wordy. 'calculate' is probably redundant.

davidcmoulton committed:
Not needed

davidcmoulton committed:
Terser

thewilkybarkid commented on src/shared-styles/_baseline.scss:
```diff
@@ -0,0 +1,18 @@
+@function size-from-units($factor) {
+  @return $_baseline-unit * $factor;
+}
+
```
```suggestion
```

thewilkybarkid committed:
Whitespace tweak

thewilkybarkid committed:
Tweak comments

thewilkybarkid committed:
Tweak names

thewilkybarkid reviewed (approved):
Couple of naming tweaks (😅), but think we're good to :shipit:.
